File: ocaml/src/nine.ml
open OUnit2

let pack xs =
  let rec go packed xs =
    match (packed, xs) with
    | ([],x::rest) -> go [[x]] rest
    | (ps,[]) -> ps
    | (p::ps,x::rest) ->
        if List.mem x p
        then go ((x :: p) :: ps) rest
        else go ([x] :: p :: ps) rest in
  List.rev (go [] xs)

let test_case =
  "pack" >::
    (fun _  ->
       assert_equal
         [["a"; "a"; "a"; "a"];
         ["b"];
         ["c"; "c"];
         ["a"; "a"];
         ["d"; "d"];
         ["e"; "e"; "e"; "e"]]
         (pack
            ["a";
            "a";
            "a";
            "a";
            "b";
            "c";
            "c";
            "a";
            "a";
            "d";
            "d";
            "e";
            "e";
            "e";
            "e"]))

File: fsharp/Nine.fs
module ``NinetyNine Problems - Problem 09``

open NUnit.Framework
open FsUnit

let pack =
  let rec aux packs xs =
    match packs, xs with
    | _, [] -> packs
    | [], x :: x_rest -> aux ([[x]]) x_rest 
    | p :: p_rest, x :: x_rest ->
      if List.contains x p then aux ((x :: p) :: p_rest) x_rest
      else aux ([x] :: p :: p_rest) x_rest
  in
    aux [] >> List.rev

[<Test>]
let ``It should make groups of adjacent equal values`` () =
  pack [1; 1; 1; 1; 2; 3; 3; 1; 1; 4; 4; 5; 5; 5; 5]
  |> should equal [
      [1; 1; 1; 1];
      [2];
      [3; 3];
      [1; 1];
      [4; 4];
      [5; 5; 5; 5]
    ]
